File: thrift/src/main/java/com/intenthq/pucket/javacompat/thrift/ThriftPucketInstantiator.java
package com.intenthq.pucket.javacompat.thrift;

import com.intenthq.pucket.Pucket;
import com.intenthq.pucket.javacompat.PucketInstantiator;
import com.intenthq.pucket.thrift.ThriftPucket;
import com.intenthq.pucket.thrift.ThriftPucketDescriptor;
import com.intenthq.pucket.util.ExceptionUtil;

import org.apache.hadoop.fs.FileSystem;
import org.apache.hadoop.fs.Path;
import org.apache.thrift.TBase;
import org.apache.thrift.TFieldIdEnum;

import java.io.IOException;

public class ThriftPucketInstantiator implements PucketInstantiator<TBase<? extends TBase<?, ?>, ? extends TFieldIdEnum>> {
  @Override
  public <T extends TBase<? extends TBase<?, ?>, ? extends TFieldIdEnum>> Pucket<T> newInstance(final Path path,
                                                                                                final FileSystem fs,
                                                                                                final String descriptor)
      throws RuntimeException, IOException {
    return ExceptionUtil.doThrow(
        ThriftPucket.findOrCreate(path,
                                  fs,
                                  ExceptionUtil.doThrow(ThriftPucketDescriptor.<T>apply(descriptor))
        )
    );
  }
}

File: thrift/src/main/scala/com/intenthq/pucket/thrift/mapreduce/ThriftPucketInstantiator.scala
package com.intenthq.pucket.thrift.mapreduce

import com.intenthq.pucket.Pucket
import com.intenthq.pucket.mapreduce.PucketInstantiator
import com.intenthq.pucket.thrift.{Thrift, ThriftPucket, ThriftPucketDescriptor}
import org.apache.hadoop.fs.{FileSystem, Path}

import scalaz.\/

/** Thrift pucket instanciator for use with output format */
class ThriftPucketInstantiator extends PucketInstantiator[Thrift] {
  /** @inheritdoc */
  def newInstance[T <: Thrift](path: Path, fs: FileSystem, descriptor: String,
                               attempts: Int = Pucket.defaultCreationAttempts,
                               retryIntervalMs: Int = Pucket.defaultRetryIntervalMs): Throwable \/ Pucket[T] =
    ThriftPucketDescriptor[T](descriptor).flatMap(ThriftPucket.findOrCreateRetry[T](path, fs, _, Pucket.defaultBlockSize, attempts, retryIntervalMs))
}

